### PR TITLE
WIP: feat: pre_hooks steps

### PR DESCRIPTION
**Pull request by w3st3ry**
This template introduces a new "preHooks" feature that triggers an action before running the step.

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report test#307.0 ✘  
* Stage 1
  * Test ✘

File: engine/step/hook.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+// Hook represents the hook data structure
+type Hook struct {
+	Name            string            `json:"name"`
+	Description     string            `json:"description""`
+	LongDescription *string           `json:"long_description,omitempty"`
+	DocLink         *string           `json:"doc_link,omitempty"`
+	Action          Executor          `json:"action"`
+	Results         map[string]string `json:"results"`
```
Comment (rbeuque74) [inline engine/step/hook.go]: I think we should change this to a map[string]interface{} to match what we have in  `models/task/task.go`

Comment (rbeuque74) [inline engine/step/hook.go]: Also, maybe we should keep this singular as `models/task/task.go`

```suggestion
	Result         map[string]interface{} `json:"result"`
```

File: engine/step/step.go [hunk 2]
```diff
@@ -264,6 +300,59 @@ func Run(st *Step, baseConfig map[string]json.RawMessage, values *values.Values,
...
+					if errors.IsBadRequest(err) {
+						st.State = StateClientError
+					} else {
+						st.State = StateServerError
+					}
+					st.Error = err.Error()
+					go noopStep(st, stepChan)
+					return
+				} else if st.ResultValidate != nil {
+					if err := st.ResultValidate(st.Output); err != nil {
```
Comment (rbeuque74) [inline engine/step/step.go]: ResultValidate is not declared on Hook structure. This code should happens on a ResultValidate structure existing inside the hook, and not on the step one.
Also, the body to validate if the one from the hook output, and not the step output.

File: engine/step/hook.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package step
+
+import (
+	"github.com/juju/errors"
+)
+
+// Hook represents the hook data structure
+type Hook struct {
+	Name            string            `json:"name"`
+	Description     string            `json:"description""`
```
Comment (rbeuque74) [inline engine/step/hook.go]:
```suggestion
	Description     string            `json:"description"`
```

Typo inside the json tag

File: engine/step/step.go [hunk 4]
```diff
@@ -264,6 +300,59 @@ func Run(st *Step, baseConfig map[string]json.RawMessage, values *values.Values,
...
+			for _, hookName := range st.PreHooks {
+				h, err := getHook(hookName)
+				if err != nil {
+					st.Error = fmt.Sprintf("get hook %s failed: %v", h.Name, err)
+					st.State = StateFatalError
+					go noopStep(st, stepChan)
+					return
+				}
+
+				o, m, err := runner.Exec(h.Name, baseCfgRaw, h.Action.Configuration, ctx)
```
Comment (fsamin) [inline engine/step/step.go]: where does this runner is coming from ? I'm afraid, it comes from `runner, err := getRunner(st.Action.Type)` so it wouldn't be the runner needed for the hooks' actions.

**Comment by ovh-cds** [comment]
CDS Report test#317.0 ✘  
* Stage 1
  * Test ✘

**Comment by ovh-cds** [comment]
CDS Report test#318.0 ✘  
* Stage 1
  * Test ✘

**Comment by rbeuque74** [comment]
We will address this another way.
Closing it for now.

**Pull request closed without merging**